File: runtime/dart/core/duration.ts
export interface DurationProps {
    days?: number;
    hours?: number;
    minutes?: number;
    seconds?: number;
    milliseconds?: number;
    microseconds?: number;
}

const durationDefaultProps = {
    days: 0,
    hours: 0,
    minutes: 0,
    seconds: 0,
    milliseconds: 0,
    microseconds: 0,
};

declare const dart: {
    core: {
        duration: (
            this: void,
            duration: Duration,
            props: DurationProps
        ) => Duration;
    };
};

export class Duration {
    private readonly _dart_getInDays: () => number = undefined as any;
    private readonly _dart_getInHours: () => number = undefined as any;
    private readonly _dart_getInMinutes: () => number = undefined as any;
    private readonly _dart_getInSeconds: () => number = undefined as any;
    private readonly _dart_getInMilliseconds: () => number = undefined as any;
    private readonly _dart_getInMicroseconds: () => number = undefined as any;
    private readonly _dart_compareTo: (
        other: Duration
    ) => number = undefined as any;
    private readonly _dart_toString: () => string = undefined as any;
    private readonly _dart_getIsNegative: () => boolean = undefined as any;
    private readonly _dart_abs: () => Duration = undefined as any;

    public getInDays() {
        return this._dart_getInDays();
    }

    public getInHours() {
        return this._dart_getInHours();
    }

    public getInMinutes() {
        return this._dart_getInMinutes();
    }

    public getInSeconds() {
        return this._dart_getInSeconds();
    }

    public getInMilliseconds() {
        return this._dart_getInMilliseconds();
    }

    public getInMicroseconds() {
        return this._dart_getInMicroseconds();
    }

    public compareTo(other: Duration) {
        return this._dart_compareTo(other);
    }

    public toString() {
        return this._dart_toString();
    }

    public getIsNegative() {
        return this._dart_getIsNegative();
    }

    public abs() {
        return this._dart_abs();
    }

    public constructor(props: DurationProps) {
        dart.core.duration(this, {
            ...durationDefaultProps,
            ...props,
        });
    }
}

File: lib/cfr/builtins/libs/dart/core/duration.dart
import 'package:hydro_sdk/cfr/builtins/boxing/boxers.dart';
import 'package:hydro_sdk/cfr/builtins/boxing/boxes.dart';
import 'package:hydro_sdk/cfr/vm/closure.dart';
import 'package:hydro_sdk/cfr/vm/context.dart';
import 'package:hydro_sdk/cfr/vm/table.dart';
import 'package:hydro_sdk/hydroState.dart';

class VMManagedDuration extends Duration implements Box<Duration> {
  final HydroTable? table;
  final HydroState hydroState;

  VMManagedDuration({
    required this.table,
    required this.hydroState,
    required int days,
    required int hours,
    required int minutes,
    required int seconds,
    required int milliseconds,
    required int microseconds,
  }) : super(
            days: days,
            hours: hours,
            minutes: minutes,
            seconds: seconds,
            milliseconds: milliseconds,
            microseconds: microseconds) {
    table!["vmObject"] = vmObject;
    table!["unwrap"] = makeLuaDartFunc(func: (List<dynamic> args) {
      return [unwrap()];
    });
    table!["_dart_getInDays"] = makeLuaDartFunc(func: (List<dynamic> args) {
      return [inDays];
    });
    table!["_dart_getInHours"] = makeLuaDartFunc(func: (List<dynamic> args) {
      return [inHours];
    });
    table!["_dart_getInMinutes"] = makeLuaDartFunc(func: (List<dynamic> args) {
      return [inMinutes];
    });
    table!["_dart_getInSeconds"] = makeLuaDartFunc(func: (List<dynamic> args) {
      return [inSeconds];
    });
    table!["_dart_getInMilliseconds"] =
        makeLuaDartFunc(func: (List<dynamic> args) {
      return [inMilliseconds];
    });
    table!["_dart_getInMicroseconds"] =
        makeLuaDartFunc(func: (List<dynamic> args) {
      return [inMicroseconds];
    });
    table!["_dart_compareTo"] = makeLuaDartFunc(func: (List<dynamic> args) {
      return [super.compareTo(args[1])];
    });
    table!["_dart_toString"] = makeLuaDartFunc(func: (List<dynamic> args) {
      return [super.toString()];
    });
    table!["_dart_getIsNegative"] = makeLuaDartFunc(func: (List<dynamic> args) {
      return [isNegative];
    });
    table!["_dart_abs"] = makeLuaDartFunc(func: (List<dynamic> args) {
      return [abs];
    });

    @override
    int? compareTo(Duration other) {
      Closure closure = table!["compareTo"];
      return closure.dispatch([
        table,
        maybeBoxObject<Duration>(
          table: HydroTable(),
          object: other,
          hydroState: hydroState,
        ),
      ], parentState: hydroState)[0];
    }

    @override
    String? toString() {
      Closure closure = table!["toString"];
      return closure.dispatch([table], parentState: hydroState)[0];
    }
  }

  Duration unwrap() => this;
  Duration get vmObject => this;
}

void loadDuration({
  required HydroState hydroState,
  required HydroTable table,
}) {
  table["duration"] = makeLuaDartFunc(func: (List<dynamic> args) {
    return [
      VMManagedDuration(
        table: args[0],
        hydroState: hydroState,
        days: args[1]["days"],
        hours: args[1]["hours"],
        minutes: args[1]["minutes"],
        seconds: args[1]["seconds"],
        milliseconds: args[1]["milliseconds"],
        microseconds: args[1]["microseconds"],
      )
    ];
  });
}
